cfg_mgmt/azure.py

```diff
@@ -0,0 +1,183 @@
+import copy
+import dataclasses
+import datetime
+import typing
+
+import msal
+import requests
+
+import cfg_mgmt
+import model
+import model.azure
+
+from cfg_mgmt.model import CfgQueueEntry
+
+AZURE_APP_ROOT_URL = 'https://graph.microsoft.com/v1.0/applications'
+
+
+@dataclasses.dataclass
+class PasswordCredential:
+    displayName: str
+    keyId: str
+    endDateTime: str
+    startDateTime: typing.Optional[str]
+    customKeyIdentifier: typing.Optional[str]
+    hint: typing.Optional[str]
+    secretText: typing.Optional[str]  # the actual password, if present
+
+    def __post_init__(self):
+        self.endDateTime = _to_datetime(self.endDateTime)
+        self.startDateTime = _to_datetime(self.startDateTime)
+
+
+def _get_access_token_for_principal(
+    service_principal: model.azure.AzureServicePrincipal,
+):
+
+    app = msal.ConfidentialClientApplication(
+        client_id=service_principal.client_id(),
+        authority=f'https://login.microsoftonline.com/{service_principal.tenant_id()}',
+        client_credential=service_principal.client_secret(),
+    )
+    scopes = ['https://graph.microsoft.com/.default',]
+    response = app.acquire_token_for_client(scopes=scopes)
+
+    if 'access_token' in response:
+        access_token = response['access_token']
+    elif 'error' in response:
+        raise RuntimeError(
+            f'Error authenticating with Microsoft Graph: {response.get("error_description")}'
+        )
+    else:
+        raise RuntimeError(f'Unexpected response: {response}')
+
+    return access_token
+
+
+def _to_datetime(
+    azureDateString:str,
+):
+    # Azure ISO date format has to be converted manually to datetime
+    pos = azureDateString.find('.')
+    if pos > 0:
+        azureDateString = azureDateString[:pos] + 'Z'
+
+    return datetime.datetime.strptime(azureDateString, '%Y-%m-%dT%H:%M:%SZ')
+
+
+def _get_passwords(
+    service_principal: model.azure.AzureServicePrincipal,
+    access_token: str,
+) -> typing.Iterable[PasswordCredential]:
+    response = requests.get(
+        url=f'{AZURE_APP_ROOT_URL}/{service_principal.object_id()}',
+        headers={'Authorization': f'Bearer {access_token}'},
+    )
+    response_dict = response.json()
+    if not 'passwordCredentials' in response_dict:
+        # no keys
+        return []
+
+    return [PasswordCredential(**e) for e in response_dict['passwordCredentials']]
```

Comment:
preferably use `dacite` - this tends to be more robust (e.g. will not break upon extra attrs)